### PR TITLE
[Merged by Bors] - port: Algebra.Order.Group.Abs

### PR DESCRIPTION
**Pull request by kim-em**
a95b16cbade0f938fc24abd05412bde1e84bab9b

- [x] depends on: #895 

Issues:
* typeclass inference
* the absolute value notation can't cope with nesting or leading minus signs.

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#895
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#895~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Algebra/Order/Group/Abs.lean [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+#align abs_neg abs_neg
+
+theorem eq_or_eq_neg_of_abs_eq {a b : α} (h : |a| = b) : a = b ∨ a = -b := by
+  simpa only [← h, eq_comm, eq_neg_iff_eq_neg] using abs_choice a
+#align eq_or_eq_neg_of_abs_eq eq_or_eq_neg_of_abs_eq
+
+theorem abs_eq_abs {a b : α} : |a| = |b| ↔ a = b ∨ a = -b := by
+  refine' ⟨fun h => _, fun h => _⟩
+  ·
+    obtain rfl | rfl := eq_or_eq_neg_of_abs_eq h <;>
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Group/Abs.lean:90]:
```suggestion
  · obtain rfl | rfl := eq_or_eq_neg_of_abs_eq h <;>
```

File: Mathlib/Algebra/Order/Group/Abs.lean [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+#align abs_pos_of_pos abs_pos_of_pos
+
+theorem abs_pos_of_neg (h : a < 0) : 0 < |a| :=
+  abs_pos.2 h.ne
+#align abs_pos_of_neg abs_pos_of_neg
+
+theorem neg_abs_le_self (a : α) : -|a| ≤ a := by
+  cases' le_total 0 a with h h
+  ·
+    calc
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Group/Abs.lean:149]:
```suggestion
  · calc
```

File: Mathlib/Algebra/Order/Group/Abs.lean [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+      -|a| = -a := congr_arg Neg.neg (abs_of_nonneg h)
+      _ ≤ 0 := neg_nonpos.mpr h
+      _ ≤ a := h
+
+  ·
+    calc
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Group/Abs.lean:155]:
```suggestion
  · calc
```

File: Mathlib/Algebra/Order/Group/Abs.lean [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+#align apply_abs_le_mul_of_one_le' apply_abs_le_mul_of_one_le'
+
+@[to_additive]
+theorem apply_abs_le_mul_of_one_le {β : Type _} [MulOneClass β] [Preorder β]
+    [CovariantClass β β (· * ·) (· ≤ ·)] [CovariantClass β β (swap (· * ·)) (· ≤ ·)] {f : α → β}
+    (h : ∀ x, 1 ≤ f x) (a : α) : f (|a|) ≤ f a * f (-a) :=
+  apply_abs_le_mul_of_one_le' (h _) (h _)
+#align apply_abs_le_mul_of_one_le apply_abs_le_mul_of_one_le
+
+/-- The **triangle inequality** in `linear_ordered_add_comm_group`s.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Group/Abs.lean:265]:
```suggestion
/-- The **triangle inequality** in `LinearOrderedAddCommGroup`s.
```

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/Algebra/Order/Group/Abs.lean [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+end CovariantAddLe
+
+section LinearOrderedAddCommGroup
+
+variable [LinearOrderedAddCommGroup α] {a b c d : α}
+
+-- Porting note:
+-- Lean can perfectly well find this instance,
+-- but in the rewrites below it is going looking for it without having fixed `α`.
+example : CovariantClass α α (swap fun x x_1 ↦ x + x_1) fun x x_1 ↦ x ≤ x_1 := inferInstance
```
Comment (jcommelin) [inline Mathlib/Algebra/Order/Group/Abs.lean:231]:
```suggestion
example : CovariantClass α α (swap fun x y ↦ x + y) fun x y ↦ x ≤ y := inferInstance
```

File: Mathlib/Algebra/Order/Group/Abs.lean [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+
+@[to_additive]
+theorem apply_abs_le_mul_of_one_le {β : Type _} [MulOneClass β] [Preorder β]
+    [CovariantClass β β (· * ·) (· ≤ ·)] [CovariantClass β β (swap (· * ·)) (· ≤ ·)] {f : α → β}
+    (h : ∀ x, 1 ≤ f x) (a : α) : f (|a|) ≤ f a * f (-a) :=
+  apply_abs_le_mul_of_one_le' (h _) (h _)
+#align apply_abs_le_mul_of_one_le apply_abs_le_mul_of_one_le
+
+/-- The **triangle inequality** in `LinearOrderedAddCommGroup`s.
+-/
```
Comment (jcommelin) [inline Mathlib/Algebra/Order/Group/Abs.lean:263]:
```suggestion
/-- The **triangle inequality** in `LinearOrderedAddCommGroup`s. -/
```

**Comment by jcommelin** [comment]
LGTM

bors d+

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**